src/core/ext/filters/client_channel/resolver/xds/xds_resolver.cc

```diff
@@ -1177,50 +1196,37 @@ void RegisterXdsResolver(CoreConfiguration::Builder* builder) {
       std::make_unique<XdsResolverFactory>());
 }
 
-void XdsClusterMap::Put(RefCountedPtr<ClusterState> state) {
-  absl::string_view cluster_name = state->cluster();
-  clusters_.emplace(cluster_name, std::move(state));
-}
-
-absl::optional<std::pair<absl::string_view, RefCountedPtr<ClusterState>>>
-XdsClusterMap::Find(absl::string_view name) const {
+RefCountedPtr<ClusterState> XdsClusterMap::Find(absl::string_view name) const {
   auto it = clusters_.find(name);
   if (it == clusters_.end()) {
-    return absl::nullopt;
+    return nullptr;
   }
-  return *it;
+  return it->second;
 }
 
 ClusterState::ClusterState(RefCountedPtr<XdsResolver> resolver,
-                           const std::string& cluster_name)
-    : resolver_(std::move(resolver)),
-      it_(resolver_->Emplace(cluster_name, WeakRef())) {}
+                           absl::string_view cluster_name)
+    : resolver_(std::move(resolver)), cluster_name_(cluster_name) {}
 
 void ClusterState::Orphan() {
-  auto* resolver = resolver_.get();
-  resolver->Run([resolver = std::move(
-                     resolver_)]() { resolver->MaybeRemoveUnusedClusters(); },
-                DEBUG_LOCATION);
+  resolver_->MaybeRemoveUnusedClustersInWorkSerializer();
```

Comment:
I just noticed that this change means that we're no longer unreffing the resolver from inside the `WorkSerializer`, which I think we need to be doing.

I suggest changing `MaybeRemoveUnusedClustersInWorkSerializer()` to take a `RefCountedPtr<XdsResolver>` as an argument, which it will `std::move()` into the lambda, so that we're no longer holding a ref here in `ClusterState`.  This means that `MaybeRemoveUnusedClustersInWorkSerializer()` can also be a static method.

Reply:
Done. Note that this method needs access to work_serializer_ so it can't be static (or the work_serializer_ needs to be made static.
I added releasing the resolver to the Orphan method.

Reply:
It doesn't fix the problem to just reset `cluster_` in this function.  The `ClusterState` object may be unreffed from outside the `WorkSerializer`, which is why we're hopping into the `WorkSerializer` to begin with, so we can't unref the resolver here.

The actual goal here is to ensure that the resolver is destroyed inside the `WorkSerializer`, and any ref that we're holding to the resolver might be the last one, which means that we always need to release those refs inside the `WorkSerializer`.  Keep in mind that the lambda that we pass to the `WorkSerializer` may run in another thread, and it's entirely possible that that other thread will finish executing it (thus releasing the ref it's holding) before this thread releases its ref, so this could very well be the last ref to the resolver.

Static methods of a class *do* have access to the class's private members; they just need a pointer to an object of that class, because they aren't called on an object of the class (i.e., they have no `this`).  So what we need to do in `XdsResolver` is this:

```
  static void MaybeRemoveUnusedClustersInWorkSerializer(
      RefCountedPtr<XdsResolver> resolver) {
    auto* resolver_ptr = resolver.get();
    resolver_ptr->work_serializer_->Run(
        [resolver = std::move(resolver)]() {
          resolver->MaybeRemoveUnusedClusters();
        },
        DEBUG_LOCATION);
  }
```

Then this method becomes:

```
void ClusterState::Orphan() {
  XdsResolver::MaybeRemoveUnusedClustersInWorkSerializer(
      std::move(resolver_));
}
```